### ISSUE TITLE
Zero Length File Uploads

### ISSUE THREAD

**Issue by nayos-scott** [dotnet/AspNetCore.Docs#19773]
This page has a couple code samples where it suggests silently skipping 0 length files, e.g.

```
    foreach (var formFile in files)
    {
        if (formFile.Length > 0)
        {
            var filePath = Path.GetTempFileName();

            using (var stream = System.IO.File.Create(filePath))
            {
                await formFile.CopyToAsync(stream);
            }
        }
    }
```

It would be nice if the documentation could mention or explain under what conditions one should expect to see 0 length files. Should this only happen when the file is actually 0 length or can this be indicative of a problem with the file upload?

From my limited experience, I have seen IFormFile intermittently report 0 length for non-empty files as reported by client side telemetry.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9a5cfa38-0e6c-4346-dafc-84ea5c513fcf
* Version Independent ID: c6787f0f-3e9a-6efb-0e6e-5ccb957d3f4a
* Content: [Upload files in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-3.1)
* Content Source: [aspnetcore/mvc/models/file-uploads.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/models/file-uploads.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Hello @nayos-scott ... It's probably arbitrary here ... I don't think that it's a guardrail against the following calls throwing (but sometimes that is why such a check is shown). I think in this case it just avoids the calls in the block if the file is zero length. I don't think we have any particular guidance for that particular check.

WRT zero-length files in general, that's up to you how to handle that scenario. We don't have any particular guidance for that scenario. One could prevent the upload and warn the user or allow it with or without an indication to the user depending on the app's requirements.

The problem with calling something out on this subject is that its an edge case in an *ocean of edge cases* here, where what we have is an advanced scenario topic with numerous possibilities for how files might be processed. For these sorts of topics, we often leave it to the dev to further research the subject and rely on multiple sources, Microsoft and external sources, to write their code.

A good bet if you want to discuss it further is to hit up the devs on the usual support channels to see what they do and what they've run into. I'm sure you know about SO, but there's a Slack Team for ASP.NET Core and a Gitter chat ...

* [ASP.NET Core Slack](http://tattoocoder.com/aspnet-slack-sign-up/)
* [ASP.NET Gitter](https://gitter.im/aspnet/Home)

**Comment by nayos-scott** [comment]
Hi @guardrex, I agree on there being an ocean of edges cases, especially with files, and outlining each one would make these docs very overwhelming and hard to consume.

In my particular case this seems like I hit a client side bug in iOS Safari as https://stackoverflow.com/questions/60729546/ios-safari-13-3-1-uploaded-file-size-is-0 sounds exactly like what I observed but the root cause is not particuarly important here.

In an ideal world it would be great if developers further researched every snippet but that's not always the case in practice. In my case, coming from other languages to try C# + dotnet for the first time, I take the official docs as gospel.

For most search queries along the lines of "how do to upload files in dotnet core", people will land on this page. When both of the first code C# snippets show this, it easy to interpret this as this is the _official_ way to handle file uploads.

This particular snippet or variants of it will then proliferate through SO, blogs, tutorials, code bases, etc.
https://www.google.com/search?q=%22formFile.Length+%3E+0%22&oq=%22formFile.Length+%3E+0%22

I think this snippet is harmful to show. It's hard to imagine a scenario where it would be desirable to silently ignore empty files. If anything, it would be better to remove these checks in the docs, so if people do encounter empty files then they get an exception, are forced to properly investigate, then implement a check that fits their use case.

**Comment by guardrex** [comment]
> I think this snippet is harmful to show. It's hard to imagine a scenario where it would be desirable to silently ignore empty files.

* If by "harm" you mean "security risk," there's no security risk to my knowledge. Also, this was approved by Microsoft's top ASP.NET Core security guru, Barry Dorrans. I'm sure he would've said something on review if he had a problem with it.
* If by "harm" you just mean as a programming practice, It's not a bad or unusual practice to avoid calls that no-op, and that code shouldn't throw if the dev doesn't want to check. Even if it did throw ... for any reason ... it would be trapped as a handled or unhandled exception at the developer's discretion.

If you want to add a bullet to that first occurrence (the bullet list of what the example does) that says something like ...

> \* Checks \`formFile.Length\` to avoid no-op calls when the file content is empty (optional).

... you're welcome to submit a PR. Mark it `Fixes #19773` and ping me in the opening comment.

**Comment by guardrex** [comment]
oh ... and note that because this is a topic with three versions of the doc in it that the list appears **three times**. That line would need to be added in **three spots**.
